sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs

```diff
@@ -0,0 +1,347 @@
+                }
+                AadAccess = acquireNewAADToken();
+            }
+
+            if (AadAccess.Expiration < DateTime.UtcNow.AddMinutes(LATENCY_SAFETY))
+            {
+                throw new Exception("The newly provided AAD token is expired.");
+            }
+
+            // Obtain a new refresh Token using the regenerated / previously valid token
+            AcrRefresh = MakeToken(authClient.GetAcrRefreshTokenFromExchangeAsync("access_token", this.LoginUrl, Tenant, null, AadAccess.TokenStr).GetAwaiter().GetResult().RefreshTokenProperty);
+
+        }
+
+        /* Acquires the required scope for a specific operation. This will be done by obtaining a chllange and parsing out the scope
+         from the ww-Authenticate header. In the event of failure (Some endpoints do not seem to return the scope) it will attempt
+         resolution through a small local resolver. */
+        public string getScope(string operation, string method, string path)
+        {
+
+            if (AcrScopes.ContainsKey(method + ">" + operation))
+            {
+                return AcrScopes[method + ">" + operation];
+            }
+
+            HttpClient runtimeClient = new HttpClient();
+            HttpResponseMessage response = null;
+            string scope;
+            try
+            {
+                response = runtimeClient.SendAsync(new HttpRequestMessage(new HttpMethod(method), operation)).GetAwaiter().GetResult();
+                Dictionary<string, string> data = parseHeader(response.Headers.GetValues("Www-Authenticate").FirstOrDefault());
+                scope = data.ContainsKey("scope") ? data["scope"] : hardcodedScopes(path);
+                AcrScopes[method + ">" + operation] = scope;
+            }
+            catch (Exception e)
+            {
+                throw new Exception("Could not identify appropiate Token scope: " + e.Message);
+            }
+            return scope;
+
+        }
+
+        /*Local resolver for endpoints that will often retuen no scope.*/
+        private string hardcodedScopes(string operation)
+        {
+            switch (operation)
+            {
+                case "/acr/v1/_catalog":
+                case "/v2/":
+                    return "registry:catalog:*";
+                default:
+                    throw new Exception("Could not determine appropiate scope for the specified operation");
+
+            }
+        }
+
+        /* Meant to parse out comma separated key value pairs delineated by the = sign. Accepts strings of the
+         format "key=value,key2=value2..." . Note this method is meant to provide limited functionality and
+         is not very robust. */
+        private Dictionary<string, string> parseHeader(string header)
+        {
+
+            Dictionary<string, string> parsed = new Dictionary<string, string>();
+
+            Regex re = new Regex(@"([\w\s]+)=""[^""]+"""); //Regex is required to allow multiple scopes like push,pull
```

Comment:
>Regex re = new Regex(@"([\w\s]+)=""[^""]+""") [](start = 12, length = 45)

Define this as a static Regex outside the scope of the method and possibly compile at initialization time.